Trapped User Flow

mindgraffiti opened the pull request:
Fixes #828 

⚠️ I'm not sure if this "fix" is a good idea. 

The reason the user gets trapped in the login flow is because they have a store, but Jetpack has returned an error of some kind, so they are unable to continue. Because real users' site configurations are unique, we have no way of knowing that Jetpack is throwing an error until they contact support. Then we can investigate the configuration.

~The store picker has so many variations and different states, I may have introduced new bugs. For example, there are times when the store picker view is loaded and no navigation controller is present. Since the user can now leave the trapped state, we have to create a nav controller and push them back to the beginning of the login flow.~ The STR have been identified and this bug has been addressed.

## To Test

~I can't consistently get the Store Picker VC into a state where the nav controller doesn't exist, so I'm unsure how to test that use case. (Sorry).~ // done!

**You'll need logins with these states:**
- a valid WP.com login with no store (empty state)
- a valid login with one valid store (no checkmark next to the single store)
- a valid login with multiple valid stores (checkmark next to store)
- a valid login that contains multiple stores, where one of the stores is invalid for some reason - such as a store that is using WooCommerce 3.4, for example. (the continue button should disable)

**Log In**
1. Be in a logged out state. Choose to log in via any of the 3 flows: email connected to WP.com, Google Sign in, or sign in by Site URL, site username, site password. None of the main flow was tampered with, only the Epilogue (the Store Picker) has changed.
2. After successfully going through the Authenticator login flow, you should land on the Store Picker screen.
3. There should not be a nav bar at the top.
4. There should be a "Try another account" button and a "Continue" button.
5. If you are testing on an iPhone X or other large screen size, the "Continue" button should be within the safe area bounds on the bottom. 

**Switch Store**
1. Navigate to My store > Settings > Switch Store
2. The "Try another account" button should not exist
3. The "Continue" button should toggle to enabled / disabled, depending on if the selected Store is valid or not.
4. The "Dismiss" button on the top left corner of the nav bar should toggle to disabled / enabled, depending on if the selected Store is valid or not. 

Update release notes:

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

jleandroperez commented on WooCommerce/Classes/Authentication/Epilogue/StorePickerViewController.swift:
```diff
@@ -362,13 +399,20 @@ private extension StorePickerViewController {
+
+        guard let navigationController = navigationController else {
+            let navController = UINavigationController(nibName: nil, bundle: nil)
+            navController.setViewControllers([loginViewController], animated: true)
```
We're not doing anything with `navController` here. Maybe dead code.. or a missing commit?

mindgraffiti replied:
setViewControllers sets the root view controller to the loginViewController and pops the StorePickerVC. Unless I have that wrong? I updated it from creating a new navController, setting a rootVC to the loginVC, and then dismissing this VC.

jleandroperez replied:
Inside the guard, a new UINavigationController is instantiated, `setViewControllers` is called (which is great!). But the new instance you've got in **navControlller** is lost, immediatelly right after the return in L413.

jaclync replied:
`navController` here is a new `UINavigationController` instance, its view controllers are set but it doesn't seem to be presented or used in the navigation stack after this line. I think in this else case, we can just return (and maybe `asserttionFailure` before)?

mindgraffiti replied:
Yep, you are correct. This isn't where I should be trying to restore navigation. The code to check for a nil navigationController was already here, hinting at a problem that couldn't be found previously. I found where the issue was, fixed that and added an `assertFailure()` as suggested here.

To reproduce the bug, check out commit `eac673c` and try this login flow:

1. Log in to a valid WP.com account that doesn't have any WC Stores. 
2. The user kills the app and re-runs it. (Or Stop Xcode and re-run, same thing.) 
3. The app delegate restores the current state, and the Store Picker screen displays again. 

Only this time, the Store Picker is set as the root view controller, which is a problem because we want to pop to root-- which should be the login view controller.

jleandroperez commented on WooCommerce/Classes/Authentication/Epilogue/StorePickerViewController.swift:
```diff
@@ -62,6 +62,9 @@ class StorePickerViewController: UIViewController {
             actionBackgroundView.layer.shadowOpacity = StorePickerConstants.backgroundShadowOpacity
+
+            // Shift the shadow so it doesn't appear on the bottom of the view.
+            actionBackgroundView.layer.shadowOffset = CGSize(width: 0, height: -5)
```
Suggestion! `StorePickerConstants.shadowOffset` perhaps?

mindgraffiti commented:
@jaclync thanks for the review!

I'll fix the 3 bugs noted in the review. Thank you for the screenshots 😀 

>Do we always want "Try another account" on the login epilogue store picker? 

This is modeled after WPiOS log in design. I did try changing the "Continue" button to "Try another account" when it errors, but it was very hack-y. I add an error case to the `state` enum and update it in the StorePicker's `displaySiteWCRequirementWarningIfNeeded` function. I can definitely go back and try it again though, if @kyleaparker wants a the continue button to change to "Try another account" instead.

kyleaparker commented:
>I can definitely go back and try it again though, if @kyleaparker wants a the continue button to change to "Try another account" instead.

The continue button should be the one in purple, is this is how it currently is?

mindgraffiti commented:
@kyleaparker @jaclync here are the updated UI fixes.

This is what the Login screen looks like when a user has a valid WP.com account, but doesn't have a store:

<img src="https://user-images.githubusercontent.com/1062444/64891211-55116d80-d63f-11e9-8819-7ab506c14df4.png" width="350" />

kyleaparker commented:
@mindgraffiti in this case do we need the continue button? I don't think it's needed if they have no stores

jaclync commented on WooCommerce/Classes/Authentication/Epilogue/StorePickerViewController.swift:
```diff
@@ -155,6 +167,8 @@ class StorePickerViewController: UIViewController {
         if configuration == .login {
             startListeningToNotifications()
+        } else if configuration == .switchingStores {
+            secondaryActionButton.isHidden = true
```
would a switch statement be good here now that there is more than one case?

mindgraffiti commented:
Latest changes:

- fixed bug where the table was overlapped by the action buttons view
- fixed a bug where iPhone 8 was showing too much whitespace (autolayout - bottom constraint adjusted and ignores safe area)

ctarda commented on WooCommerce/Classes/Authentication/Epilogue/StorePickerViewController.swift:
```diff
@@ -362,13 +398,18 @@ private extension StorePickerViewController {
         let loginViewController = ServiceLocator.authenticationManager.loginForWordPressDotCom()
+
+        guard let navigationController = navigationController else {
+            assertionFailure("Navigation error: one of the login / logout states is not correctly handling navigation. No navigationController found.")
```
I like this, at least we'd get warned when testing locally that we are in an inconsistent state

ctarda commented on WooCommerce/Classes/Authentication/Epilogue/StorePickerViewController.swift:
```diff
@@ -379,11 +420,20 @@ private extension StorePickerViewController {
             case .invalidWCVersion:
-                self?.updateActionButtonAndTableState(animating: false, enabled: false)
-                self?.displayFancyWCRequirementAlert(siteName: siteName)
+                self?.toggleDismissButton(enabled: false)
```
🤔 this is a switch within a switch... in a method that might be a bit too long already. Would you consider extracting each of the branches in the outer switch to a separate method? That might help with readability

mindgraffiti replied:
done ✔️

ctarda commented on WooCommerce/Classes/Authentication/Epilogue/StorePickerViewController.swift:
```diff
@@ -474,6 +536,7 @@ extension StorePickerViewController: UITableViewDataSource {
 
     func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
         guard let site = state.site(at: indexPath) else {
+            hideActionButton()
```
This is very nice. I know, sometimes, it seems overkill to extract a method with just two lines, but now, when reading this code, it is telling us the intent, we don't need to figure it out by gathering pieces of code here and there.

loremattei commented:
Hey! I'm moving this to 2.8 since 2.7 has been cut. Please, if you want this PR to make it to 2.7, feel free to move it back, target the release branch and ping me to build a new beta.

mindgraffiti commented:
Thank you @loremattei! 🙇‍♀

mindgraffiti commented:
Broke apart the method that had a switch nested inside of a switch. Will merge after Hound is done.
